third_party/move/move-compiler-v2/src/pipeline/reference_safety_processor.rs

```diff
@@ -0,0 +1,1453 @@
+                // message, and showing derived references helps. Otherwise derived
+                // references may be noisy.
+                if !alive.before.contains_key(temp) {
+                    secondary_edges.extend(state.effective_descendants(edge.target))
+                }
+            }
+        }
+        primary_edges
+            .into_iter()
+            .filter_map(|(_, e)| self.borrow_edge_info("previous ", state, only_mut, e))
+            .chain(
+                secondary_edges
+                    .into_iter()
+                    .filter_map(|(_, e)| self.borrow_edge_info("used by ", state, only_mut, e)),
+            )
+            .collect()
+    }
+
+    fn borrow_edge_info(
+        &self,
+        prefix: &str,
+        _state: &LifetimeState,
+        only_mut: bool,
+        e: &BorrowEdge,
+    ) -> Option<(Loc, String)> {
+        if e.is_mut || !only_mut {
+            if let Some(loc) = &e.loc {
+                use BorrowEdgeKind::*;
+                let mut_prefix = if e.is_mut { "mutable " } else { "" };
+                return Some((
+                    loc.clone(),
+                    format!("{}{}{}", prefix, mut_prefix, match e.kind {
+                        BorrowLocal => "local borrow",
+                        BorrowGlobal => "global borrow",
+                        BorrowField(..) => "field borrow",
+                        BorrowIndex => "index borrow",
+                        Call(..) => "call result",
+                        Skip => return None,
+                    },),
+                ));
+            }
+        }
+        None
+    }
+
+    // ---------------------------------------------------------------------------------------------
+    // Program Steps
+
+    /// Process an assign instruction. This checks whether the source is currently borrowed and
+    /// rejects a move if so. Constructs a `Skip` edge in the lifetime graph if references
+    /// are assigned.
+    fn assign(
+        &self,
+        state: &mut LifetimeState,
+        code_offset: CodeOffset,
+        id: AttrId,
+        dest: TempIndex,
+        src: TempIndex,
+        kind: AssignKind,
+        alive: &LiveVarInfoAtCodeOffset,
+    ) {
+        // Check validness
+        let mode = if kind == AssignKind::Move {
+            ReadMode::Move
+        } else {
+            ReadMode::Copy
+        };
+        self.check_read_local(state, id, src, mode, alive);
+        self.check_write_local(state, id, dest, alive);
+        let ty = self.ty(src);
+        if ty.is_reference() {
+            // Track reference in the graph as a Skip edge.
+            let loc = self.loc(id);
+            let label = state.make_local(src, code_offset, 0);
+            let descendant = state.replace_local(dest, code_offset, 1);
+            state.add_edge(
+                label,
+                BorrowEdge::new(
+                    BorrowEdgeKind::Skip,
+                    ty.is_mutable_reference(),
```

Comment:
what is the meaning of the `is_mut` field for skip edges?

Reply:
It just inherits the mutability, here we use the type to get the info.